### PR TITLE
Make the core model/state of the search (SearchParameter) immutable (for real)

### PR DESCRIPTION
**Pull request by bobylito**
The point of this pull request is to provide real immutability for the state of the search that is encapsulated into the helper (`helper.state`)

In practice, this will help to easily test that the previous state is different from the current by comparing references. It will also help test properties for the future, you can now easily modify the model without having to revert the changes.

```javascript
var state0 = SearchParameters.make( {
  disjunctiveFacets : [ "facet" ]
} );

var state1 = state0.addDisjunctiveRefinement( "facet", "value" );

state0 !== state1 //true
deepEqual( state0, state1 ); // false

//actually

state0.disjunctiveRefinements !== state1.disjunctiveRefinements;
```

### CODE REVIEW

File: src/SearchParameters/RefinementList.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    if( lib.isRefined( refinementList, attribute, value ) ) {
+      return lib.removeRefinement( refinementList, attribute, value );
+    }
+
+    return lib.addRefinement( refinementList, attribute, value );
+  },
+  /**
+   * Clear all or parts of a RefinementList. Depending on the arguments, three
+   * behaviors can happen :
+   *  - if not attribute is provided : clears the whole list
```
Comment (pixelastic) [inline src/SearchParameters/RefinementList.js]: s/not/no/

File: src/SearchParameters/index.js [hunk 2]
```diff
@@ -331,9 +354,11 @@ SearchParameters.prototype = {
    * @return {SearchParameters}
    */
   setHitsPerPage : function setHitsPerPage( n ) {
-    return this.mutateMe( function( m ) {
-      m.hitsPerPage = n;
-      m.page = 0;
+    if( this.HitsPerPage === n ) return this;
```
Comment (pixelastic) [inline src/SearchParameters/index.js]: Shouldn't this be `hitsPerPage` (lowercase) ?

File: src/SearchParameters/index.js [hunk 3]
```diff
@@ -344,9 +369,11 @@ SearchParameters.prototype = {
    * @return {SearchParameters}
    */
   setTypoTolerance : function setTypoTolerance( s ) {
```
Comment (pixelastic) [inline src/SearchParameters/index.js]: Can't we use a more explicit argument name, like in the in `setPage` or `setQuery`, like `typoTolerance` ?

**Comment by pixelastic** [comment]
LGTM :+1:

**Comment by vvo** [comment]
✅ reviewed

**Comment by bobylito** [comment]
Thanks @vvo @pixelastic

**Pull request merged**